File: 03_patterns/swift/mvvm_clean_arc/Clean/Data/DataSource/TaskLocalDataSource.swift
import Foundation

final class TaskLocalDataSource {
    private let store: SharedTaskStore

    init(store: SharedTaskStore = .shared) {
        self.store = store
    }

    func fetchTasks() -> [Task] {
        store.fetchTasks().map(\.asTask)
    }

    func saveTask(title: String) -> [Task] {
        store.createTask(title: title).map(\.asTask)
    }

    func removeTask(id: UUID) -> [Task] {
        store.deleteTask(id: id).map(\.asTask)
    }
}

private extension SharedTaskRecord {
    var asTask: Task {
        Task(id: id, title: title)
    }
}

File: 03_patterns/flutter/lib/clean/data/data_source/task_local_data_source.dart
import '../../../shared/storage/shared_task_store.dart';
import '../../domain/entities/task.dart';

class TaskLocalDataSource {
  TaskLocalDataSource({SharedTaskStore? store})
      : _store = store ?? SharedTaskStore.shared;

  final SharedTaskStore _store;

  List<Task> fetchTasks() {
    return _store.fetchTasks().map(_asTask).toList();
  }

  List<Task> saveTask({required String title}) {
    return _store.createTask(title: title).map(_asTask).toList();
  }

  List<Task> removeTask({required String id}) {
    return _store.deleteTask(id: id).map(_asTask).toList();
  }

  Task _asTask(SharedTaskRecord record) {
    return Task(id: record.id, title: record.title);
  }
}
